### PR TITLE
MRG: Allow preload=False for readers

### PR DESCRIPTION
**Pull request by Eric89GXL**
Currently BTi and EGI must be preloaded, this PR will fix that.

Only BTi implemented currently. @dengemann feel free to have a look at that bit since it tweaks code you've worked with.

### CODE REVIEW

**Comment by agramfort** [comment]
@dengemann <https://github.com/dengemann> this one if for you.

**Comment by Eric89GXL** [comment]
EGI done.

This seems weird, but we could actually do the same thing for `RawArray`. For example if you have 32-bit (or 16-bit) data, it would store the original array, and only upconvert it when necessary. But it's probably overkill. So ready for review/merge from my end.

As far as I can tell, the EGI reader is not tested as extensively for data equivalence as our other readers. This is probably due to not having the MNE-C equivalent `mne_egi2fiff` function. @teonlamont or @dengemann it looks like wrote the EGI code -- do you have other software e.g. in MATLAB where you could save out the resulting data array, so we could compare to a known correct result? It would make modifications to the code more solid/reliable/believable than they are now.

File: mne/io/egi/egi.py [hunk 1]
```diff
@@ -65,41 +65,24 @@ def my_fread(*x, **y):
...
-                       6: ('>f8', 8)}[info['precision']]
-
-    info.update({'dtype': dtype, 'bytesize': bytesize})
-    beg_dat = fid.tell()
-
-    for ii in range(info['n_events']):
-        fid.seek(beg_dat + (int(n_channels) + ii) * bytesize, 0)
-        events[ii] = np.fromfile(fid, dtype, n_samples)
-        fid.seek(int((n_channels + n_events) * bytesize), 1)
-    return events
```
Comment (Eric89GXL) [inline mne/io/egi/egi.py]: FYI this bit of event reading was 1) never used in practice and 2) incorrect

File: mne/io/egi/egi.py [hunk 2]
```diff
@@ -242,28 +236,23 @@ def __init__(self, input_fname, montage=None, eog=None, misc=None,
...
             event_ids = np.arange(len(include_)) + 1
-            try:
-                logger.info('    Synthesizing trigger channel "STI 014" ...')
-                logger.info('    Excluding events {%s} ...' %
-                            ", ".join([k for i, k in enumerate(event_codes)
-                                       if i not in include_]))
-                new_trigger = _combine_triggers(egi_events[include_],
-                                                remapping=event_ids)
-                data = np.concatenate([data, new_trigger])
-            except RuntimeError:
```
Comment (Eric89GXL) [inline mne/io/egi/egi.py]: remove unnecessary `try/except`

**Comment by Eric89GXL** [comment]
@dengemann would this also make it so that you don't need the `pdf_fname=None` functionality anymore, since BTi data is no longer forced to be preloaded? In other words, can you use the public `read_raw_bti` function for your purposes instead of the private info reading function? It would be nice to get rid of the `sfreq=None` bits of code in `meas_info.py` that we had to add for that special case.

File: mne/io/egi/tests/test_egi.py [hunk 3]
```diff
@@ -24,17 +24,19 @@ def test_io_egi():
...
+        assert_true(w[1].category == RuntimeWarning)
         msg = 'Did not find any event code with more than one event.'
-        assert_true(msg in '%s' % w[0].message)
+        assert_true(msg in '%s' % w[1].message)
 
     include = ['TRSP', 'XXX1']
-    raw = _test_raw_reader(read_raw_egi, False, True, input_fname=egi_fname,
-                           include=include)
+    with warnings.catch_warnings(record=True):  # preload=None
+        raw = _test_raw_reader(read_raw_egi, True, False,
```
Comment (jaeilepp) [inline mne/io/egi/tests/test_egi.py]: Why did you change ``test_blocks`` to False?

File: mne/io/bti/tests/test_bti.py [hunk 4]
```diff
@@ -49,20 +51,14 @@ def test_read_config():
...
-
-
 def test_crop_append():
     """ Test crop and append raw """
-    raw = _test_raw_reader(read_raw_bti, False, True, pdf_fname=pdf_fnames[0],
-                           config_fname=config_fnames[0],
-                           head_shape_fname=hs_fnames[0])
+    with warnings.catch_warnings(record=True):  # preload warning
+        warnings.simplefilter('always')
+        raw = _test_raw_reader(read_raw_bti, True, test_blocks=False,
```
Comment (jaeilepp) [inline mne/io/bti/tests/test_bti.py]: Same here. Why ``test_blocks=False``?

File: mne/io/egi/egi.py (hunk 2 above)
Comment (dengemann) [inline mne/io/egi/egi.py]: why was it unnecessary? are corner cases still handled with the changes?

File: mne/io/egi/egi.py (hunk 1 above)
Comment (dengemann) [inline mne/io/egi/egi.py]: can you be more specific? How wasn't it uses, how was it wrong?

**Comment by dengemann** [comment]
> @dengemann would this also make it so that you don't need the pdf_fname=None functionality anymore, since BTi data is no longer forced to be preloaded? 

The problem there was that the info is somehow distributed across `config` and `pdf` files. It is more difficult to extract a single bit of information e.g. the sfreq from the 'pdf' file which includes also the data. It was the only way I saw to handle it without digging too deep into the data reader, which is a mess. In principle it would be nice to have what you propose, it will just require some editing to the lowest level of BTI readers. You will somehow need to keep the config and the pdf around and you need to memorize pointer positions to get that done. The file format is a bit annoying, headers and data segments are positioned alternatingly, you will need to read the next bit of data to get to the next header, and so on.

As to non-formal EGI testing, practically we used this function a lot with hundreds of datasets. Probably our testing is maybe limited as we mainly saw files from one site.

**Comment by Eric89GXL** [comment]
> It is more difficult to extract a single bit of information e.g. the sfreq from the 'pdf' file which includes also the data

At this point, it should automatically be doing that, I think. Since data isn't preloaded, it should be fast to get measurement info. So I guess the question is: should `preload=False` cover your use case sufficiently that you won't need to use the private info reading function? If so I'll work on refactoring the code.

> As to non-formal EGI testing, practically we used this function a lot with hundreds of datasets. Probably our testing is maybe limited as we mainly saw files from one site.

It's good that you're confident the reader was working properly before. The problem with the approach of testing-in-practice like this instead of testing-in-software-suite is that when PRs like this come up to refactor or improve code, we can't have confidence in the results without a repeat of this practical testing. That's why I'm looking for e.g. a `.npz` or `.mat` file with the data loaded from a different program to validate against. This is similar to how we validate against `mne_*2fiff` outputs for other readers. But of course it would also be good if you could test the data in practice on top of that.

File: mne/io/egi/tests/test_egi.py (hunk 3 above)
Comment (Eric89GXL) [inline mne/io/egi/tests/test_egi.py]: These readers don't really have a notion of "blocks", but I guess I could leave it in to test slicing. But then why have it be an option at all? If it's to save time, I'll update the docstring to talk about testing slicing, since right now it refers to needing to test `buffer_size_sec`.

File: mne/io/egi/egi.py (hunk 2 above)
Comment (Eric89GXL) [inline mne/io/egi/egi.py]: If you look at the old `_combine_triggers` code there was one place it explicitly raised `RuntimeError` and I have it return `None` now. If there were other places it implicitly did it, then we should refactor it anyway because it's nasty to try to figure out of from the code, but I don't think there were.

Comment (Eric89GXL) [inline mne/io/egi/egi.py]: This is one reason I strongly prefer using `try/except` in a limited manner for only simple cases, because in cases like this it's hard to tell under what conditions the "right" error could actually occur from reading the code.

File: mne/io/egi/egi.py [hunk 5]
```diff
@@ -108,7 +91,7 @@ def _combine_triggers(data, remapping=None):
     first = np.nonzero(data[0])[0]
     for d in data[1:]:
         if np.intersect1d(d.nonzero()[0], first).any():
-            raise RuntimeError('Events must be mutually exclusive')
+            return
```
Comment (Eric89GXL) [inline mne/io/egi/egi.py]: @dengemann this is where the old `raise` used to be

**Comment by Eric89GXL** [comment]
Okay I managed to convert an EGI file using EEGLAB. @agramfort okay to add it to the repo? It's 100kB when compressed using gzip.

**Comment by Eric89GXL** [comment]
The alternative is to add it to our `testing` dataset, which is simple enough.

**Comment by Eric89GXL** [comment]
Okay I pushed a commit with the changes removing `pdf_fname=None` now that `preload=False` can be done instead. @dengemann can you confirm that it's fast enough for your use case?

Ready for review/merge from my end. @agramfort if you want me to put the 100kB file in the testing repo I'm happy to do that and squash out the addition to the repo here.

File: mne/io/egi/egi.py (hunk 1 above)
Comment (Eric89GXL) [inline mne/io/egi/egi.py]: the returned `event` value was never used, see below for the only call to this function.

File: mne/io/egi/egi.py [hunk 6]
```diff
@@ -192,22 +189,16 @@ def __init__(self, input_fname, montage=None, eog=None, misc=None,
             logger.info('Reading EGI header from %s...' % input_fname)
             egi_info = _read_header(fid)
             logger.info('    Reading events ...')
-            _read_events(fid, egi_info)  # update info + jump
```
Comment (Eric89GXL) [inline mne/io/egi/egi.py]: i.e. here, output never used. And the forwarding of the file pointer was never made use of, either, since the next reading function, the old `_read_data` used a call to `fid.seek(info['bytes_per_slice'] * start, 0)` which is an absolute seek.

File: mne/io/egi/egi.py (hunk 1 above)
Comment (Eric89GXL) [inline mne/io/egi/egi.py]: It was wrong because it assumed the data was stored row-major (if you think of rows as being channels and cols as being time), i.e. with one channel, then the next, then the next. But looking at the old `_read_data` function, which reads data properly, the data are stored column-major, i.e. all the channels for the first time sample, then all the channels for the next time sample, etc.

File: mne/io/egi/egi.py [hunk 7]
```diff
@@ -65,79 +65,62 @@ def my_fread(*x, **y):
...
     fid.seek(36 + info['n_events'] * 4, 0)  # skip header
-    readsize = (info['n_channels'] + info['n_events']) * info['n_samples']
-    final_shape = (info['n_samples'], info['n_channels'] + info['n_events'])
-    data = np.fromfile(fid, info['dtype'], readsize).reshape(final_shape).T
-    return data
+    for si in range(info['n_samples']):
+        # skip data channels
+        fid.seek(info['n_channels'] * info['bytesize'], 1)
+        # read event channels
+        events[:, si] = np.fromfile(fid, info['dtype'], info['n_events'])
```
Comment (Eric89GXL) [inline mne/io/egi/egi.py]: ... thus why I modified the `_read_events` function to loop through sample-by-sample, skipping over the channel data along the way. Should hopefully be fast enough since it shouldn't need to actually read that much data.

**Comment by Eric89GXL** [comment]
@dengemann any time to review this one? It would be good to get this reading PR in before other reading PRs mature (e.g., #2676) because it touches a bit of `base` code.

**Comment by Eric89GXL** [comment]
@dengemann did you get my ping on this one? Just wanted to make sure you noticed I pinged you on two PRs (this one and the one you merged already, #2643) in case you thought both pings were for the same PR.

File: mne/io/bti/bti.py [hunk 8]
```diff
@@ -1060,42 +1016,57 @@ def __init__(self, pdf_fname, config_fname='config',
...
                  ecg_ch='E31', eog_ch=('E63', 'E64'),
-                 verbose=None):
-
-        if pdf_fname is None:
-            raise ValueError('pdf_fname must be a path, not None')
+                 preload=None, verbose=None):
+        if preload is None:
+            warnings.warn('preload is True by default but will be changed to '
+                          'False in v0.12. Please explicitly set preload.',
+                          DeprecationWarning)
```
Comment (jaeilepp) [inline mne/io/bti/bti.py]: What's the difference with ``warnings.warn`` and ``logger.warning``? For me, this message doesn't appear when I read the file.

**Comment by jaeilepp** [comment]
Noticed something weird. Try reading the EGI test file with ``preload=False``, use ``raw.plot`` and click right arrow.

File: mne/io/bti/bti.py (hunk 8 above)
Comment (Eric89GXL) [inline mne/io/bti/bti.py]: One goes through the `warnings` module, the other through `logging`. We usually do `warnings.warn` for deprecations. I think it should appear for you if you have `preload=None` (default) when reading since we have this line in `utils.py`:
```
# force show of DeprecationWarning even on python 2.7
warnings.simplefilter('default')
```
Then I just noticed this in `decoding/base.py`:
```
            # We need deprecation warnings to always be on in order to
            # catch deprecated param values.
            # This is set in utils/__init__.py but it gets overwritten
            # when running under python3 somehow.
            warnings.simplefilter("always", DeprecationWarning)
```
Can you try replacing the line in `utils.py` with the one from decoding to see if it makes the warning appear for you? You'll probably need to restart your interpreter and re-import mne-python.

**Comment by Eric89GXL** [comment]
@jaeilepp bug reproduced, will fix

File: mne/io/bti/bti.py (hunk 8 above)
Comment (jaeilepp) [inline mne/io/bti/bti.py]: Yes, the message appears with this change.

Comment (Eric89GXL) [inline mne/io/bti/bti.py]: okay will change

Comment (jaeilepp) [inline mne/io/bti/bti.py]: Okay, it seems that this changes the state of the warning module. Now it throws the warning even if I change it back to default, but I can see from another console with the old setting that there is no warning when ``preload=None``.

Comment (Eric89GXL) [inline mne/io/bti/bti.py]: Yeah it will change it for the `warnings` module. I think we want it to, though, to be safe.

Comment (Eric89GXL) [inline mne/io/bti/bti.py]: Maybe it's overkill, though...

Comment (Eric89GXL) [inline mne/io/bti/bti.py]: Let's tackle that in a separate issue. What's here is our standard procedure, so let's figure out what's going on with your setup separately.

**Comment by Eric89GXL** [comment]
Okay @jaeilepp comments addressed, the plotting thing should work now. It caused me to look into the `_test_raw_reading` function, which I noticed wasn't testing the failing condition properly, so I changed it. Can you make sure I didn't botch your function? Time to run `nosetests mne/io` is down a few sec (149 to 145) on my machine with the changes, which is understandable since I cut out a bit of the writing tests.

**Comment by jaeilepp** [comment]
LGTM

**Comment by dengemann** [comment]
Thanks Eric and Jaakko!
